compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangParserListener.java

```diff
@@ -2648,6 +2651,22 @@ public void exitSimpleLiteral(BallerinaParser.SimpleLiteralContext ctx) {
         } else if ((node = ctx.QuotedStringLiteral()) != null) {
             String text = node.getText();
             text = text.substring(1, text.length() - 1);
+            String originalText = text; // to log the errors
+            Pattern pattern = Pattern.compile(Constants.UNICODE_REGEX);
+            Matcher matcher = pattern.matcher(text);
+            while (matcher.find()) {
+                String hexStringVal = matcher.group(1);
+                int hexDecimalVal = Integer.parseInt(hexStringVal, 16);
+                if ((hexDecimalVal >= 0xD800 && hexDecimalVal <= 0xDFFF) || hexDecimalVal > 0x10FFFF) {
+                    String hexStringWithBraces = matcher.group(0);
+                    int offset = originalText.indexOf(hexStringWithBraces) + 1;
+                    dlog.error(new DiagnosticPos(diagnosticSrc, pos.sLine, pos.eLine, pos.sCol + offset,
+                                    pos.sCol + offset + hexStringWithBraces.length()),
+                            DiagnosticCode.INVALID_UNICODE, hexStringWithBraces);
+                }
+                text = matcher.replaceFirst("\\\\u" + hexStringVal);
+                matcher = pattern.matcher(text);
+            }
```

Comment:
Instead of discarding the previous match data and starting from the beginning we can start from the previous `matchers` end index.

```
int idx = 0;
while (matcher.find(idx)) {
    ....
    text = matcher.replaceFirst("\\\\u" + hexStringVal);
    idx = matcher.end() - 2;  // 2 assuming we remove two characters only by the replace
    matcher = pattern.matcher(text);
}
```

Reply:
+1 and done